### ISSUE TITLE
Multicore working in BLAS, but not in Keras or Theano

### ISSUE THREAD

**Issue by prinsherbert** [fchollet/keras#2544]
Continuing on https://github.com/fchollet/keras/issues/2516 , with two experiments.

I noticed that even though theano can use multicore, keras in general doesn't (on my machine). That is if I run `fit`, only one core is used.

I've made a minimal working example of theano using multiple cores (it has a speedup around 9x w.r.t. no `os.environ` settings and 1 core):

```
import os
os.environ['MKL_NUM_THREADS'] = '8'
os.environ['GOTO_NUM_THREADS'] = '8'
os.environ['OMP_NUM_THREADS'] = '8'
os.environ['THEANO_FLAGS'] = 'device=cpu,blas.ldflags=-lblas -lgfortran'

import numpy
import theano
import theano.tensor as T

M=2000
N=2000
K=2000
iters=100
order='C'

a = theano.shared(numpy.ones((M, N), dtype=theano.config.floatX, order=order))
b = theano.shared(numpy.ones((N, K), dtype=theano.config.floatX, order=order))
c = theano.shared(numpy.ones((M, K), dtype=theano.config.floatX, order=order))
f = theano.function([], updates=[(c, 0.4 * c + .8 * T.dot(a, b))])

for i in range(iters):
    f(y)
```

And altered this script for a Keras experiment which does not use multiple cores (even though I want it so badly):

```
import os
os.environ['MKL_NUM_THREADS'] = '8'
os.environ['GOTO_NUM_THREADS'] = '8'
os.environ['OMP_NUM_THREADS'] = '8'
os.environ['THEANO_FLAGS'] = 'device=cpu,blas.ldflags=-lblas -lgfortran'

import numpy
from keras.models import Sequential
from keras.layers import Dense

coeffs = numpy.random.randn(100)

x = numpy.random.randn(100000, 100);
y = numpy.dot(x, coeffs) + numpy.random.randn(100000) * 0.01

model = Sequential()
model.add(Dense(20, input_shape=(100,)))
model.add(Dense(1, input_shape=(20,)))
model.compile(optimizer='rmsprop', loss='categorical_crossentropy')

model.fit(x, y, verbose=0, nb_epoch=10)
```

The output `Using Theano backend.` indicates that theano was used as a backend, yet only one core was used during the entire run of the script. I get that the BLAS-implementation is responsible for the use of multiple cores, but what I do not get is why `fit` doesn't apply multiple cores when the theano experiment shows that it is possible?

Why is `fit` not executed on multiple cores, even though `0.4 * c + .8 * T.dot(a, b)` is?

**Comment by prinsherbert** [comment]
http://stackoverflow.com/questions/36908978/keras-not-using-multiple-cores

**Comment by prinsherbert** [comment]
I might have concluded that Keras was at fault to soon, this simple theano example also does not run multi-core:

    import os
    os.environ['MKL_NUM_THREADS'] = '8'
    os.environ['GOTO_NUM_THREADS'] = '8'
    os.environ['OMP_NUM_THREADS'] = '8'
    os.environ['THEANO_FLAGS'] = 'device=cpu,blas.ldflags=-lblas -lgfortran'

    import numpy
    import theano
    import theano.tensor as T

    M=2000
    N=2000
    K=2000
    iters=100
    order='C'

    coeffs = numpy.random.randn(100)
    x = numpy.random.randn(100000, 100).astype(theano.config.floatX)
    y = (numpy.dot(x, coeffs) + numpy.random.randn(100000) * 0.01).astype(theano.config.floatX).reshape(100000, 1)

    x_shared = theano.shared(x)
    y_shared = theano.shared(y)

    x_tensor = T.matrix('x')
    y_tensor = T.matrix('y')

    W0_values = numpy.asarray(
        numpy.random.uniform(
            low=-numpy.sqrt(6. / 120),
            high=numpy.sqrt(6. / 120),
            size=(100, 20)
        ),
        dtype=theano.config.floatX
    )
    W0 = theano.shared(value=W0_values, name='W0', borrow=True)

    b0_values = numpy.zeros((20,), dtype=theano.config.floatX)
    b0 = theano.shared(value=b0_values, name='b0', borrow=True)

    output0 = T.dot(x_tensor, W0) + b0

    W1_values = numpy.asarray(
        numpy.random.uniform(
            low=-numpy.sqrt(6. / 120),
            high=numpy.sqrt(6. / 120),
            size=(20, 1)
        ),
        dtype=theano.config.floatX
    )
    W1 = theano.shared(value=W1_values, name='W1', borrow=True)

    b1_values = numpy.zeros((1,), dtype=theano.config.floatX)
    b1 = theano.shared(value=b1_values, name='b1', borrow=True)

    output1 = T.dot(output0, W1) + b1

    params = [W0, b0, W1, b1]
    cost = ((output1 - y_tensor) ** 2).sum()

    gradients = [T.grad(cost, param) for param in params]

    learning_rate = 0.0000001

    updates = [
        (param, param - learning_rate * gradient)
        for param, gradient in zip(params, gradients)
    ]

    train_model = theano.function(
        inputs=[],#x_tensor, y_tensor],
        outputs=cost,
        updates=updates,
        givens={
            x_tensor: x_shared,
            y_tensor: y_shared
        }
    )

    errors = []
    for i in range(1000):
        errors.append(train_model())
        
    print(errors[0:50:])

**Comment by prinsherbert** [comment]
It seems that my beef was with `openmp_elemwise_minsize`, it was set too high for any of the operations (in mini-batch mode) to be parellalized. The depressing part is that there's not really a speed up if I set it to 1 or 500 (200000 is the default)

**Comment by nouiz** [comment]
Threads on CPU are costly, so you need tensor of significant size before it help.

**Comment by prinsherbert** [comment]
@nouiz I was expecting the `openmp_elemwise_minsize` setting is not there for no reason. But maybe it's a good idea to develop the architecture to supersede openmp_elemwise_minsize for most / all operations.

**Comment by nouiz** [comment]
I don't understand what you tell. How do you want to supersede that? This
just change the threashold where we enable the openmp code we have for
elemwise operation. We won't reimplement an OpenMP back-end if that is what
you suggest. If not, what do you mean?

**Comment by nouiz** [comment]
Theano do not run in parallel node of the graph on the CPU. We should have
this for the GPU this summer.

On the CPU, we only parallelist those operations:

http://deeplearning.net/software/theano/tutorial/multi_cores.html

**Comment by prinsherbert** [comment]
With architecture I meant the configuration of the neural network (layers / nodes / etc), such that it can learn more within fewer epochs by having more nodes (hence bigger operations and more parallelization).

**Comment by nouiz** [comment]
Theano currently paralellize inside one node, not between nodes in the
graph. So currently if you want more parallization, you most do the
opposite.

**Comment by prinsherbert** [comment]
@nouiz Well, changing the architecture could also mean changing dimensions in a node, which gives the CPU more work (if dimensions get bigger), but also more to parallelize and hopefully a model that better fits the problem (possibly with weight decay / ridge / ... techniques to battle overfitting).

**Comment by prinsherbert** [comment]
I tried this experiment again, with a simple convex regression problem, in both keras and theano using different settings. This is the code:

For Keras:

    import os, sys
    os.environ['MKL_NUM_THREADS'] = sys.argv[1]
    os.environ['GOTO_NUM_THREADS'] = sys.argv[1]
    os.environ['OMP_NUM_THREADS'] = sys.argv[1]
    os.environ['THEANO_FLAGS'] = sys.argv[2]

    import numpy, theano, time
    from keras.models import Model
    from keras.layers import Input, Dense
    from keras.optimizers import SGD, Optimizer
    import keras.backend as K

    class GD(Optimizer):
        '''Gradient descent
        # Arguments
            lr: float >= 0. Learning rate.
        '''
        def __init__(self, lr=0.01, *args, **kwargs):
            super(GD, self).__init__(**kwargs)
            self.__dict__.update(locals())
            self.iterations = K.variable(0.)
            self.lr = K.variable(lr)

        def get_updates(self, params, constraints, loss):
            self.updates = [
                (self.iterations, self.iterations + 1.)
            ] + [
                (p, p - self.lr * g)
                for p, g in zip(params, self.get_gradients(loss, params))
            ]
            return self.updates

        def get_config(self):
            return {"name": self.__class__.__name__, "lr": float(K.get_value(self.lr))}

    M=2000
    N=500
    P=2000 # K
    iters=30
    order='C'

    X  = numpy.array(numpy.random.randn(M, P), dtype=theano.config.floatX, order=order)
    W0 = numpy.array(numpy.random.randn(P, N), dtype=theano.config.floatX, order=order)
    Y  = numpy.dot(X, W0)

    l0 = Input(shape = (P,), name='input')
    l2 = Dense(N, activation='linear', name='output', bias=False)(l0)

    #optimizer = GD(lr=1)
    #optimizer = SGD(lr=0.1, momentum=0.9, nesterov=True)
    optimizer = SGD(lr=1)
    loss = {'output': 'mse'}

    model = Model(l0, l2)
    model.compile(optimizer=optimizer, loss=loss)

    #grace iteration, to make sure everything is compiled and ready
    model.fit( X, Y, nb_epoch=1, verbose=0 )

    t0 = time.time()
    out = model.fit( X, Y, nb_epoch=iters, verbose=0 )
    print( time.time() - t0 )

    print( numpy.mean((W0 - model.layers[1].W.get_value()) ** 2) )

For Theano:

    import os, sys, time
    os.environ['MKL_NUM_THREADS'] = sys.argv[1]
    os.environ['GOTO_NUM_THREADS'] = sys.argv[1]
    os.environ['OMP_NUM_THREADS'] = sys.argv[1]
    os.environ['THEANO_FLAGS'] = sys.argv[2]

    import numpy
    import theano
    import theano.tensor as T

    M=2000
    N=500
    K=2000
    iters=30
    order='C'

    X  = numpy.array( numpy.random.randn(M, K), dtype=theano.config.floatX, order=order )
    W0 = numpy.array( numpy.random.randn(K, N), dtype=theano.config.floatX, order=order )
    Y  = numpy.dot( X, W0 )

    Xs = theano.shared( X, name='x' )
    Ys = theano.shared( Y, name='y' )
    Ws = theano.shared( numpy.array(numpy.random.randn(K, N) / (K + N), dtype=theano.config.floatX, order=order), name='w' )

    cost = T.sum( (T.dot(Xs, Ws) - Ys) ** 2)

    gradient = theano.grad(cost, Ws)

    f = theano.function([], theano.shared(0), updates = [(Ws, Ws - 0.0001 * gradient)])

    #grace iteration, to make sure everything is compiled and ready
    f()

    t0 = time.time()
    for i in range(iters):
        f()
    print( time.time() - t0 )

    print( numpy.mean((W0 - Ws.get_value()) ** 2) )

These are the results:

    training time (sec)  | no openmp | 1 thread | 2 threads | 3 threads | 4 threads | 5 threads | 6 threads | 7 threads | 8 threads |
    Theano               | 10.252196 | 3.459806 | 1.9559076 | 1.4941468 | 1.1789379 | 1.4890902 | 1.4941354 | 1.4441833 | 1.4668899 |
   Keras GD + 1 batch   | 10.366756 | 3.760082 | 2.2454292 | 1.7523956 | 1.4521133 | 1.7854723 | 1.8001911 | 1.7941901 | 1.9129860 |
    Keras GD             | 13.999158 | 6.705660 | 4.3801815 | 3.7460491 | 2.8626194 | 3.8962366 | 3.9913094 | 3.6463823 | 3.6867950 |
    Keras SGD            | 18.239614 | 11.97003 | 8.3076971 | 7.4235661 | 6.8767440 | 7.9793632 | 8.1491499 | 8.0962055 | 8.0962844 |
    Keras SGD + Momentum | 19.649117 | 13.27763 | 9.8552392 | 8.7199593 | 8.0262494 | 9.4668932 | 9.0756802 | 9.4254622 | 10.488802 |

no openmp means: `argv == [..., '' 'device=cpu']`, n threads means `argv == [..., '$n', 'device=cpu,blas.ldflags=-lblas -lgfortran']`. The number of threads used was consistent with what I saw in the task manager.

It seems as if Keras is slower for the minibatch-feature and non-GD optmizers, which is what one can expect.

Anyway, this simple network reacts on the #cores as theano does, which is good :)
